tools/fpgabist/bist/bist_afu.c

```diff
@@ -0,0 +1,387 @@
+	//				   CSR_DST_ADDR);
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_NUM_LINES, LPBK1_BUFFER_SIZE / CL(1));
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_NUM_LINES");
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CFG, 0x42000);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	status_ptr = dsm_ptr + DSM_STATUS_TEST_COMPLETE/8;
+
+	/* Start the test */
+	res = fpgaWriteMMIO32(accelerator_handle, 0, CSR_CTL, 3);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_CFG");
+
+	/* Wait for test completion */
+	while (0 == ((*status_ptr) & 0x1)) {
+		usleep(100);
+	}
+
+
+/************* Begin BIST *************/
+        /* Perform BIST Check */
+	printf("Running BIST Test\n");
+        uint64_t data = 0;
+        double count = 0;
+        unsigned int bist_mask = ENABLE_DDRA_BIST;
+        unsigned int ddra_bist_result = 0;
+
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
+
+        fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_CTRL_ADDR,&data);
+	while((CHECK_BIT(data,27) != 0x08000000)){
+	  printf("Enable Test: reading result #%f: %04x\n", count, (unsigned int)data);
+	  if (count >= MAX_COUNT){
+		fprintf(stderr, "BIST not enabled!\n");
+                return -1;
+          }
+	  count++;
+	}
+	printf("BIST is enabled.  Reading status register\n");
+
+	count = 0;
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,9) != 0x200) && (CHECK_BIT(data,8) != 0x100) && (CHECK_BIT(data,7) != 0x80) && 
+            (CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
+          fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
+                 &data); 
+          if (count >= MAX_COUNT){
+		fprintf(stderr, "DDR Bank A BIST Timed Out.\n");
+                break;
+          }
+          count++;
+	  usleep(100);
+        }
+
+        if (CHECK_BIT(data,9) == DDRA_BIST_PASS) {
+                ddra_bist_result = DDRA_BIST_PASS;
+        } else if (CHECK_BIT(data,8) == DDRA_BIST_FAIL) {
+                ddra_bist_result = DDRA_BIST_FAIL;
+        } else if (CHECK_BIT(data,7) == DDRA_BIST_TIMEOUT) {
+                ddra_bist_result = DDRA_BIST_TIMEOUT;
+        } else {
+                ddra_bist_result = DDRA_BIST_FATAL_ERROR;
+        }
+
+        
+        bist_mask = ENABLE_DDRB_BIST;
+        count = 0;
+	res = fpgaWriteMMIO32(accelerator_handle, 0, DDR_BIST_CTRL_ADDR, bist_mask);
+	ON_ERR_GOTO(res, out_free_output, "writing CSR_BIST");
+        while ((CHECK_BIT(data,10) != 0x400) && (CHECK_BIT(data,11) != 0x800) && (CHECK_BIT(data,12) != 0x1000)){
+          fpgaReadMMIO64(accelerator_handle, 0, DDR_BIST_STATUS_ADDR,
+                 &data); 
+          if (count >= MAX_COUNT){
+		fprintf(stderr, "DDR Bank B BIST Timed Out.\n");
+                break;
+          }
+          count++;
+	  usleep(100);
+        }
+
+        if (ddra_bist_result == DDRA_BIST_PASS) {
```

Comment:
Why are you doing this? Instead of assigning new value `ddra_bist_result` just move this logic to lines L290-298 and you'll have the same thing.